### PR TITLE
Check amd64 extension support at startup

### PR DESCRIPTION
**Pull request by TheNumbat**
Native executables now check CPUID support at startup. For example, a binary compiled with `-fbmi2` will abort at startup when run on a CPU that does not support BMI2.

Each compilation unit defines a weak symbol for each architecture flag that was enabled in its compiler invocation. The linker aggregates all weak symbols, so the runtime (`caml_assert_ach_extensions`) can check which symbols were included in the final executable.

Testing this is a bit difficult, as the result of any test would depend on the CPU it runs on. So far I've manually tested that nothing currently triggers the check, and that enabling `-fprefetchwt1` triggers the proper abort (with and without -internal-assembler).

Concerns:
- The macOS assember (nasm) doesn't support weak symbols, so they are not emitted on macOS.
- `x86_masm.ml` asserts on weak symbols, so they're not emitted on (non-mingw/cygwin) Windows either.
- The code is duplicated between runtime4/ and runtime/.

### CODE REVIEW

**Review by gretay-js — COMMENTED** [review]
- Supporting weak symbols in the internal assembler deserves its own separate PR, which includes a small test, if possible. It may be tricky to add a test for `.weak` - the only thing that emits `.weak` at the moment is `%probe ~enabled_at_init:true`  For now, I haven't reviewed the `internal_assembler` and `x86_binary_emitter` changes in detail.  

- I assume the copies of isa.c and isa.h are identical, so read only one of each.
- 
- I think we need a compilation flag  (and the corresponding OCAMLPARAM) to disable cpuid checking generation at compile time, and/or an OCAMLRUNPARAM to disable it on a particular run of an executable, if it's not too complicated to add. 

- Shared linking / dynlink can be fixed as a follow-up PR.

- resolving weak symbols across archives seems to be fine in the normal case and it was confirmed by a small test.  @poechsel pointed out that if at least one compilation unit in a library is live, it will be extracted and the symbols will be linked into the executable and all compilation units use the same arch extensions. otherwise, the library won't be linked. He also mentioned that a corner case (wrapped symbols) might not work correctly with older linkers but I don't think it is a concern for this hack.

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -922,6 +922,39 @@ let emit_push_trap_label handler =
...
+    match system with
+    | S_macosx | S_win32 | S_win64 -> true
+    | _ -> false
+
+  let emit_symbols () =
+    if not (is_nasm_or_masm ()) then
+    List.iter (fun ext ->
+      let name = symbol ext in
+      D.label name;
+      D.weak name
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/amd64/emit.mlp:952]: I was confused about unresolved weak symbols, we do need the `D.label` here, as discussed offline with @TheNumbat, who will add a comment with a bit more documentation about our assumptions.

File: backend/amd64/emit.mlp [hunk 2]
```diff
@@ -1963,6 +1996,9 @@ let begin_assembly unix =
     D.extrn "caml_raise_exn" NEAR;
   end;
 
+  D.data ();
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:1993]: Moved to the end

File: ocaml/runtime4/isa.c [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+#define CAML_INTERNALS
+
+#include "caml/misc.h"
+#include "caml/mlvalues.h"
+#include "caml/isa.h"
+
+// Weak symbols are only supported in ELF
+
+#if (defined __x86_64__ || defined _M_X64) && defined __ELF__
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply ocaml/runtime4/isa.c:23]: Yes - removed

File: ocaml/runtime4/isa.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+#define PCLMULQDQ_BIT    (1 << 1)
+#define SSSE3_BIT        (1 << 9)
+#define SSSE4_1_BIT      (1 << 19)
+#define SSSE4_2_BIT      (1 << 20)
+#define POPCNT_BIT       (1 << 23)
+
+// CPUID with EAX = 0x80000001, result in ECX
+
+#define ABM_BIT             (1 << 5)
+#define _3DNOWPREFFETCH_BIT (1 << 8)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply ocaml/runtime4/isa.c:65]: Yeah `ABM` and `3DNOW` are the original AMD names for the extensions. Changed to LZCNT/PRETETCHW for consistency. 

I think keeping LZCNT under BMI is fine though - there are no CPUs where we would want to enable LZCNT but not BMI. It's actually redundant to check LZCNT_BIT on top of BMI_BIT, so maybe we should just delete the check.

File: tests/simd/stubs.c [hunk 5]
```diff
@@ -1,9 +1,12 @@
 
 #include <caml/memory.h>
 #include <caml/simd.h>
+#include <caml/isa.h>
 #include <smmintrin.h>
 #include <assert.h>
 
+CAML_ARCH_SSE4_2;
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply tests/simd/stubs.c]: I've removed these in this PR; I'll make another to add checking in C stubs.

**Review by gretay-js — APPROVED** [review]

**Pull request merged**